common/src/main/java/com/discordsrv/common/feature/onlinesync/OnlineSyncModule.java

```diff
@@ -0,0 +1,239 @@
+        DiscordRole role = discordSRV.discordAPI().getRoleById(config.roleId);
+        if (role == null) {
+            return Task.completed(OnlineSyncResult.ROLE_DOESNT_EXIST);
+        }
+
+        if (!role.getGuild().getSelfMember().canInteract(role)) {
+            return Task.completed(OnlineSyncResult.ROLE_CANNOT_INTERACT);
+        }
+
+        return role.getGuild().retrieveMemberById(userId)
+                .then(member -> Boolean.TRUE.equals(newState)
+                        ? member.addRole(role).thenApply(v -> (ISyncResult) GenericSyncResults.ADD_DISCORD)
+                        : member.removeRole(role).thenApply(v -> GenericSyncResults.REMOVE_DISCORD))
+                .mapException(RestErrorResponseException.class, t -> {
+                    if (t.getErrorCode() == ErrorResponse.UNKNOWN_MEMBER.getCode()) {
+                        throw new SyncFail(OnlineSyncResult.NOT_A_GUILD_MEMBER);
+                    }
+                    if (t.getErrorCode() == ErrorResponse.MISSING_PERMISSIONS.getCode()) {
+                        throw new SyncFail(GenericSyncResults.MEMBER_CANNOT_INTERACT);
+                    }
+                    throw t;
+                });
+    }
+
+    @Override
+    protected Task<ISyncResult> applyGame(OnlineSyncConfig.Entry config, UUID playerUUID, @Nullable Boolean newState) {
+        return Task.completed(GenericSyncResults.WRONG_DIRECTION);
+    }
+
+    @Override
+    public void onPlayerConnected(PlayerConnectedEvent event) {
+        resyncAll(OnlineSyncCause.PLAYER_JOINED_SERVER, Someone.of(event.player().uniqueId()));
+    }
+
+    @Subscribe
+    public void onPlayerChangeWorld(PlayerChangedWorldEvent event) {
+        resyncAll(OnlineSyncCause.PLAYER_CHANGED_WORLD, Someone.of(event.player().uniqueId()));
+    }
+
+    @Subscribe
+    public void onPlayerDisconnected(PlayerDisconnectedEvent event) {
+        resyncAll(OnlineSyncCause.PLAYER_LEFT_SERVER, Someone.of(event.player().uniqueId()));
+    }
+
+    @Subscribe
+    public void onShutdown(DiscordSRVShuttingDownEvent event) {
+        JDA jda = discordSRV.jda();
+        if (jda == null) {
+            return;
+        }
+
+        HashMap<Member, ArrayList<Role>> pairs = new HashMap<>();
+        for (OnlineSyncConfig.Entry config : configs()) {
+            DiscordRole role = discordSRV.discordAPI().getRoleById(config.roleId);
+            if (role == null) {
+                return;
+            }
+
+            role.getGuild().asJDA().getMembersWithRoles(role.asJDA()).forEach(member ->
+                    pairs.computeIfAbsent(member, __ -> new ArrayList<>()).add(role.asJDA())
+            );
+        }
+
+        CompletableFuture<Result<List<Void>>> future = RestAction.allOf(pairs.entrySet().stream().map(entry ->
+                entry.getKey().getGuild().modifyMemberRoles(entry.getKey(), Collections.emptyList(), entry.getValue())
+        ).collect(Collectors.toList())).mapToResult().submit();
```

Comment:
`modifyMemberRoles` cannot be used if the bot doesn't have a role higher than the member's highest role. This will therefor not work on some members while the single add/removes will